src/storage/src/source/reclock/compat.rs

```diff
@@ -305,8 +329,17 @@ impl RemapHandle for PersistHandle {
         Antichain<Self::IntoTime>,
     )> {
         if !std::mem::replace(&mut self.snapshot_produced, true) {
-            for ((update, _), ts, diff) in self
-                .read_handle
+            let mut read_handle = self
+                .persist_client
+                .open_leased_reader::<SourceData, (), mz_repr::Timestamp, Diff>(
+                    self.metadata.remap_shard.clone(),
+                    &format!("reclock snapshot reader {}", self.id),
+                )
+                .await
+                .context("error opening persist shard")
+                .expect("invalid usage");
```

Comment:
how expensive is it to keep opening and closing `ReadHandle`'s? Why can't we just hold one like before?

Reply:
what sort of frequency does this happen on? the source/sink error stuff registers and expires every time it writes to save the heartbeating, which seems correct to me. but if this happens regularly, it should be long-lived, yes

Reply:
This will be done once in the lifetime of the remap operator, so once after (re-)starting and then never again. Before, we were holding on to the read handle because we were relying on it to hold back the since. Now, we use the critical handle to hold back the since and the read handle would go completely unused.

Reply:
This `next` method is called on every `sync`, which we do on every `append_batch` (so every `tick`): https://github.com/MaterializeInc/materialize/blob/main/src/storage/src/source/reclock.rs#L619, so I think we should still hold a full `ReadHandle`

Reply:
But the code path that creates and uses the `ReadHandle` is entered only once:https://github.com/MaterializeInc/materialize/blob/96ae351f31955eb556243012e16b3d0883fe0554/src/storage/src/source/reclock/compat.rs#L328
So it seems wasteful to keep a (heartbeating) `ReadHandle` around just for that. wdyt?

Reply:
Whenever something like this happens my leaky abstraction lights start flashing. We shouldn't have to be thinking about heartbeats and cost models when we're using `persist`, we should "just use it". The thing that stands out to me here is: Why do critical handles not allow you to read the collection? The fact that they have stable identities and don't need heartbeats seems completely orthogonal to their function, which is "permission to read a shard from the handle's `since`". In that vain I would expect that both critical and non critical handles would allow you to read/subscribe to a shard, and only compact when asked to, making them behave identically with the only exception being reliance on heartbeats or stable identity. That's mostly food for thought for @danhhz.

In this particular case, even if the `persist` API stayed the same, I would still err towards easier to read and reason about code by just holding onto a `ReadHandle` forever. The load that this will generate is exactly the same load that we would have if the remap handle was handled by the controller. It's only because we had to hotfix a problem that we colocated them, therefore typing the optimization won't buy us anything as we'll undo it very soon.

Reply:
>  Why do critical handles not allow you to read the collection? The fact that they have stable identities and don't need heartbeats seems completely orthogonal to their function, which is "permission to read a shard from the handle's since". In that vain I would expect that both critical and non critical handles would allow you to read/subscribe to a shard, and only compact when asked to, making them behave identically with the only exception being reliance on heartbeats or stable identity. That's mostly food for thought for @danhhz.

Yeah, this is certainly an unresolved question in my mind. The primary reason they don't right now is that no critical reader users needed to be able to read. It sounds like that's the state we'll get back to once we allow queries on the remap collection?